modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageUnpacker.java

```diff
@@ -413,6 +433,109 @@ public Object unpackObject(int dataType) {
         throw new IgniteException("Unknown client data type: " + dataType);
     }
 
+    /**
+     * Packs an object.
+     *
+     * @return Object array.
+     * @throws IllegalStateException in case of unexpected value type.
+     */
+    public Object[] unpackObjectArray() {
+        assert refCnt > 0 : "Unpacker is closed";
+
+        if (tryUnpackNil())
+            return null;
+
+        int size = unpackInt();
+
+        if (size == 0)
+            return new Object[size];
+
+        Object[] args = new Object[size];
+
+        for (int i = 0; i < size; i++) {
+            if (tryUnpackNil())
+                continue;
+
+            ImmutableValue v = unpackValue();
+
+            switch (v.getValueType()) {
+                case BOOLEAN:
+                    args[i] = v.asBooleanValue().getBoolean();
+
+                    break;
+                case INTEGER:
+                    args[i] = extractNumberValue(v.asIntegerValue());
+
+                    break;
+                case FLOAT:
+                    args[i] = v.asFloatValue().toDouble();
+
+                    break;
+                case STRING:
+                    args[i] = v.asStringValue().asString();
+
+                    break;
+                case BINARY:
+                    args[i] = v.asBinaryValue().asByteArray();
+
+                    break;
+                case EXTENSION: {
+                    ExtensionValue val = v.asExtensionValue();
+
+                    if (val.getType() == ClientMsgPackType.UUID) {
+                        args[i] = bytesToUUID(val);
+                        continue;
+                    } else
+                        throw new IllegalStateException("Unexpected extended value type: " + val.getType());
+                }
+                default:
+                    throw new IllegalStateException("Unexpected value type: " + v.getValueType());
+            }
+        }
+        return args;
+    }
+
+    /**
+     * Extracts number value according to value range.
+     *
+     * @param iv IntegerValue.
+     * @return Numeric java object selected based on the value range.
+     */
+    private Object extractNumberValue(IntegerValue iv) {
```

Comment:
This logic may result in a weird behavior where different types are returned depending on the value. Especially if we return this to the user. This is why `unpackObject` method takes a type code - I suggest using the same approach here.

Reply:
Agree. Now I pack type of serialized value.